common/src/main/java/io/netty/util/internal/PlatformDependent0.java

```diff
@@ -541,6 +546,74 @@ public Object run() {
                 DIRECT_BUFFER_CONSTRUCTOR != null ? "available" : "unavailable");
     }
 
+    private static Method getVirtualThreadCheckMethod() {
+        if (JAVA_VERSION < 19) {
+            return null;
+        }
+        try {
+            Method isVirtualMethod = Thread.class.getMethod("isVirtual");
+            // Call once to make sure the invocation works.
+            boolean isVirtual = (Boolean) isVirtualMethod.invoke(Thread.currentThread());
+            return isVirtualMethod;
+        } catch (Throwable e) {
+            if (logger.isTraceEnabled()) {
+                logger.debug("virtual thread check method is not available: ", e);
+            } else {
+                logger.debug("virtual thread check method is not available: ", e.getMessage());
+            }
+            return null;
+        }
+    }
+
+    private static Class<?> getBaseVirtualThreadClass() {
+        if (JAVA_VERSION < 19) {
+            return null;
+        }
+        try {
+            return Class.forName("java.lang.BaseVirtualThread", false, ClassLoader.getSystemClassLoader());
```

Comment:
I believe this assume too much of the internal(s) of JDK here - I know that's better performing but still...a bit brittle!

Reply:
Yeah, the `Class.isInstance()` is about 3x faster than `Method::invoke` based on my local benchmark.
and the `Class.forName()` is an attempt to check whether it's available to use, if not, it will fall back to use `Method::invoke`.

Reply:
> I believe this assume too much of the internal(s) of JDK here

Understood your concern, the internal class `BaseVirtualThread` might be changed in the future, I done this because it's a trade off for the `invoke` speed, if we want less assumption on the JDK internal implementation, it will be fine to only use `Method::invoke`, but will have some more overhead than `Class.isInstance()`.

Reply:
> How does invokeExact fare? For 4.2 at least

I think `invokeExact` will be better than `invoke`, and should not be faster than `Class.isInstance()`, but still needs to do some benchmarks to verify.

Reply:
I think it should be enough to avoid thread-local optimizations only if the current thread is definitely virtual, which means 1) the `isVirtual` method is available, and 2) it returns `true`.

An `invokeExact` should optimize to be the same as a direct method call (maybe modulo a bit of inlining budget).

So I think we can achieve our result without jumping through as many hoops as is currently being done in this PR.

Reply:
>  1) the isVirtual method is available

The `isVirtual` may not be available for some reason, for example be stopped by a `SecurityManager`.

> An `invokeExact` should optimize to be the same as a direct method call 

Another PR https://github.com/netty/netty/pull/14951 done it using `invokeExact` for 4.2. But this PR is based on 4.1, which the `invokeExact` may not available to use.

> jumping through as many hoops as is currently being done in this PR.

The jumping hoops were trying to avoid calling the less efficient `invoke` method as possible, another PR https://github.com/netty/netty/pull/14951 using `invokeExact` has less hoops.

Reply:
> The invocation of `isVirtual` may not be available to call for some reason, for example be stopped by a `SecurityManager`.

I don't think this is a realistic concern.

I checked all openjdk versions 21 through 24, and the `isVirtual` is just an instanceof check on `BaseVirtualThread` with no `SecurityManager` involved. The security manager mechanism is also deprecated for removal, so I don't think we need to worry about it for future versions.

Reply:
Yes the security manager is deprecated, but still can be enabled with VM option `-Djava.security.manager=allow` at least for JDK-21.
A custom `SecurityManager` may still affect the reflection operations IIRC.

But it doesn't bother, after the latest commit, we now only return `true/false`, PLTA, also the 4.2-version https://github.com/netty/netty/pull/14951 as well, many thanks.

Reply:
I guess the idea of keeping the `BaseVirtualClass` instanceof check is to avoid potentially expensive reflective calls?

> A custom `SecurityManager` may still be able to affect the reflection operations IIRC.

It will be a `checkPackageAccess` for `java.lang`, which would also be done for any class loading. Sounds like a restriction that will break a lot of things. There won't be any other permission check for the member access, because the method is public, so it's accessible by default.

Reply:
> I guess the idea of keeping the BaseVirtualClass instanceof check is to avoid potentially expensive reflective calls?

Yes it is.

> It will be a checkPackageAccess for java.lang, which would also be done for any class loading. Sounds like a restriction that will break a lot of things. There won't be any other permission check for the member access, because the method is public, so it's accessible by default.

Yes, so we can assume that the `MethodHandle` of `Thread.isVirtual()` should be always obtainable and invokable if JDK >=19. That will make the logic easier to go. I will change the code soon.